File: python/src/graph_engineering/primitives/router.py
"""Deterministic policy-based route selection."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal, NotRequired, TypeAlias, TypedDict, TypeGuard, cast

from ..models import JsonObject
from ._numbers import portable_integer
from ._serialization import compact_json
from .errors import (
    PrimitiveValidationError,
    PrimitiveValidationIssue,
    PrimitiveValidationIssueCode,
)

RouteSelectionReasonCode: TypeAlias = Literal[
    "REQUESTED_ROUTES_SELECTED",
    "DEFAULT_SELECTED_NO_REQUEST",
    "DEFAULT_SELECTED_UNKNOWN_ROUTE",
    "ESCALATION_SELECTED_LOW_CONFIDENCE",
    "NO_REQUESTED_ROUTE",
    "UNKNOWN_ROUTE",
    "MULTIPLE_ROUTES_FOR_SINGLE",
    "MULTICAST_LIMIT_EXCEEDED",
]
_PolicyKind: TypeAlias = Literal["single", "multi"]
_DuplicateCode: TypeAlias = Literal["DUPLICATE_ID", "DUPLICATE_SELECTION"]


class RouteSelectionRequest(TypedDict):
    requestedRoutes: list[str]
    confidenceBasisPoints: NotRequired[int | float]


class RouteConfidencePolicy(TypedDict):
    minimumBasisPoints: int | float
    escalationRoute: str


class _RouteSelectionPolicyBase(TypedDict):
    allowedRoutes: list[str]
    defaultRoute: NotRequired[str]
    confidence: NotRequired[RouteConfidencePolicy]


class SingleRouteSelectionPolicy(_RouteSelectionPolicyBase):
    kind: Literal["single"]


class MultiRouteSelectionPolicy(_RouteSelectionPolicyBase):
    kind: Literal["multi"]
    maxMulticast: int | float


RouteSelectionPolicy: TypeAlias = (
    SingleRouteSelectionPolicy | MultiRouteSelectionPolicy
)

_SAFE_ROUTE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
_REQUEST_FIELDS = frozenset({"requestedRoutes", "confidenceBasisPoints"})
_POLICY_FIELDS = frozenset(
    {"kind", "allowedRoutes", "defaultRoute", "confidence", "maxMulticast"}
)
_CONFIDENCE_FIELDS = frozenset({"minimumBasisPoints", "escalationRoute"})


@dataclass(frozen=True, slots=True)
class RouteSelectionResult:
    """Immutable route decision with an exact TypeScript-compatible export."""

    routed: bool
    reason_code: RouteSelectionReasonCode
    requested_routes: tuple[str, ...]
    selected_routes: tuple[str, ...]
    unknown_routes: tuple[str, ...]
    confidence_basis_points: int | None
    used_default: bool
    escalated: bool

    def to_dict(self) -> JsonObject:
        return {
            "routed": self.routed,
            "reasonCode": self.reason_code,
            "requestedRoutes": list(self.requested_routes),
            "selectedRoutes": list(self.selected_routes),
            "unknownRoutes": list(self.unknown_routes),
            "confidenceBasisPoints": self.confidence_basis_points,
            "usedDefault": self.used_default,
            "escalated": self.escalated,
        }

    def to_json(self) -> str:
        return compact_json(self.to_dict())

    def to_json_bytes(self) -> bytes:
        return self.to_json().encode("utf-8")


@dataclass(frozen=True, slots=True)
class _RequestSnapshot:
    requested_routes: tuple[str, ...]
    confidence_basis_points: int | None
    confidence_present: bool


@dataclass(frozen=True, slots=True)
class _ConfidenceSnapshot:
    minimum_basis_points: int
    escalation_route: str


@dataclass(frozen=True, slots=True)
class _PolicySnapshot:
    kind: _PolicyKind
    allowed_routes: tuple[str, ...]
    default_route: str | None
    confidence: _ConfidenceSnapshot | None
    confidence_present: bool
    max_multicast: int | None


def _is_plain_record(value: object) -> TypeGuard[dict[str, object]]:
    return (
        isinstance(value, dict)
        and type(value) is dict
        and all(isinstance(key, str) and type(key) is str for key in value)
    )


def _add_issue(
    issues: list[PrimitiveValidationIssue],
    path: str,
    code: PrimitiveValidationIssueCode,
    message: str,
) -> None:
    issues.append(PrimitiveValidationIssue(path, code, message))


def _unknown_fields(
    value: dict[str, object],
    allowed: frozenset[str],
    path: str,
    issues: list[PrimitiveValidationIssue],
) -> None:
    for name in sorted(value):
        if name not in allowed:
            _add_issue(issues, f"{path}/{name}", "UNKNOWN_FIELD", "unknown field")


def _valid_route_id(value: object) -> TypeGuard[str]:
    return (
        isinstance(value, str)
        and type(value) is str
        and _SAFE_ROUTE_ID.fullmatch(value) is not None
        and value not in {".", ".."}
    )


def _validate_route_array(
    value: object,
    path: str,
    *,
    duplicate_code: _DuplicateCode,
    require_non_empty: bool,
    issues: list[PrimitiveValidationIssue],
) -> tuple[str, ...]:
    if not isinstance(value, list) or type(value) is not list:
        _add_issue(issues, path, "TYPE", "expected a route ID array")
        return ()
    if require_non_empty and not value:
        _add_issue(
            issues,
            path,
            "INVALID_POLICY",
            "allowedRoutes must not be empty",
        )

    routes: list[str] = []
    seen: set[str] = set()
    for index, raw_route in enumerate(value):
        item_path = f"{path}/{index}"
        if not _valid_route_id(raw_route):
            _add_issue(issues, item_path, "UNSAFE_ID", "expected a safe route ID")
            continue
        if raw_route in seen:
            label = "requested" if duplicate_code == "DUPLICATE_SELECTION" else "allowed"
            _add_issue(
                issues,
                item_path,
                duplicate_code,
                f"duplicate {label} route '{raw_route}'",
            )
        else:
            seen.add(raw_route)
        routes.append(raw_route)
    return tuple(routes)


def _validate_request(
    value: object,
    issues: list[PrimitiveValidationIssue],
) -> _RequestSnapshot:
    path = "#/request"
    if not _is_plain_record(value):
        _add_issue(issues, path, "TYPE", "expected a plain request object")
        return _RequestSnapshot((), None, False)
    _unknown_fields(value, _REQUEST_FIELDS, path, issues)

    requested_routes: tuple[str, ...] = ()
    if "requestedRoutes" not in value:
        _add_issue(
            issues,
            f"{path}/requestedRoutes",
            "REQUIRED",
            "requestedRoutes is required",
        )
    else:
        requested_routes = _validate_route_array(
            value["requestedRoutes"],
            f"{path}/requestedRoutes",
            duplicate_code="DUPLICATE_SELECTION",
            require_non_empty=False,
            issues=issues,
        )

    confidence_present = "confidenceBasisPoints" in value
    confidence_basis_points: int | None = None
    if confidence_present:
        confidence_basis_points = portable_integer(
            value["confidenceBasisPoints"],
            minimum=0,
            maximum=10_000,
        )
        if confidence_basis_points is None:
            _add_issue(
                issues,
                f"{path}/confidenceBasisPoints",
                "INVALID_CONFIDENCE",
                "confidenceBasisPoints must be a safe integer in [0, 10000]",
            )

    return _RequestSnapshot(
        requested_routes,
        confidence_basis_points,
        confidence_present,
    )


def _validate_confidence_policy(
    value: object,
    allowed_routes: frozenset[str],
    allowed_routes_valid: bool,
    issues: list[PrimitiveValidationIssue],
) -> _ConfidenceSnapshot | None:
    path = "#/policy/confidence"
    if not _is_plain_record(value):
        _add_issue(
            issues,
            path,
            "INVALID_POLICY",
            "expected a plain confidence policy object",
        )
        return None
    _unknown_fields(value, _CONFIDENCE_FIELDS, path, issues)

    minimum_basis_points: int | None = None
    if "minimumBasisPoints" not in value:
        _add_issue(
            issues,
            f"{path}/minimumBasisPoints",
            "REQUIRED",
            "minimumBasisPoints is required",
        )
    else:
        minimum_basis_points = portable_integer(
            value["minimumBasisPoints"],
            minimum=1,
            maximum=10_000,
        )
        if minimum_basis_points is None:
            _add_issue(
                issues,
                f"{path}/minimumBasisPoints",
                "INVALID_CONFIDENCE",
                "minimumBasisPoints must be a safe integer in [1, 10000]",
            )

    escalation_route: str | None = None
    if "escalationRoute" not in value:
        _add_issue(
            issues,
            f"{path}/escalationRoute",
            "REQUIRED",
            "escalationRoute is required",
        )
    else:
        raw_escalation = value["escalationRoute"]
        if not _valid_route_id(raw_escalation):
            _add_issue(
                issues,
                f"{path}/escalationRoute",
                "UNSAFE_ID",
                "expected a safe route ID",
            )
        else:
            escalation_route = raw_escalation
            if allowed_routes_valid and escalation_route not in allowed_routes:
                _add_issue(
                    issues,
                    f"{path}/escalationRoute",
                    "INVALID_POLICY",
                    "escalationRoute must be declared in allowedRoutes",
                )

    if minimum_basis_points is None or escalation_route is None:
        return None
    return _ConfidenceSnapshot(minimum_basis_points, escalation_route)


def _validate_policy(
    value: object,
    issues: list[PrimitiveValidationIssue],
) -> _PolicySnapshot:
    path = "#/policy"
    if not _is_plain_record(value):
        _add_issue(issues, path, "TYPE", "expected a plain policy object")
        return _PolicySnapshot("single", (), None, None, False, None)
    _unknown_fields(value, _POLICY_FIELDS, path, issues)

    kind: _PolicyKind = "single"
    kind_valid = False
    if "kind" not in value:
        _add_issue(issues, f"{path}/kind", "REQUIRED", "policy kind is required")
    else:
        raw_kind = value["kind"]
        if (
            not isinstance(raw_kind, str)
            or type(raw_kind) is not str
            or raw_kind not in {"single", "multi"}
        ):
            _add_issue(
                issues,
                f"{path}/kind",
                "INVALID_POLICY",
                "kind must be 'single' or 'multi'",
            )
        else:
            kind = cast(_PolicyKind, raw_kind)
            kind_valid = True

    allowed_routes: tuple[str, ...] = ()
    allowed_issues_start = len(issues)
    allowed_present = "allowedRoutes" in value
    if not allowed_present:
        _add_issue(
            issues,
            f"{path}/allowedRoutes",
            "REQUIRED",
            "allowedRoutes is required",
        )
    else:
        allowed_routes = _validate_route_array(
            value["allowedRoutes"],
            f"{path}/allowedRoutes",
            duplicate_code="DUPLICATE_ID",
            require_non_empty=True,
            issues=issues,
        )
    allowed_routes_valid = len(issues) == allowed_issues_start and allowed_present
    allowed_set = frozenset(allowed_routes)

    default_route: str | None = None
    if "defaultRoute" in value:
        raw_default = value["defaultRoute"]
        if not _valid_route_id(raw_default):
            _add_issue(
                issues,
                f"{path}/defaultRoute",
                "UNSAFE_ID",
                "expected a safe route ID",
            )
        else:
            default_route = raw_default
            if allowed_routes_valid and default_route not in allowed_set:
                _add_issue(
                    issues,
                    f"{path}/defaultRoute",
                    "INVALID_POLICY",
                    "defaultRoute must be declared in allowedRoutes",
                )

    confidence_present = "confidence" in value
    confidence: _ConfidenceSnapshot | None = None
    confidence_config_valid = False
    if confidence_present:
        confidence_issues_start = len(issues)
        confidence = _validate_confidence_policy(
            value["confidence"],
            allowed_set,
            allowed_routes_valid,
            issues,
        )
        confidence_config_valid = (
            confidence is not None and len(issues) == confidence_issues_start
        )

    max_multicast: int | None = None
    max_present = "maxMulticast" in value
    if kind_valid and kind == "single":
        if max_present:
            _add_issue(
                issues,
                f"{path}/maxMulticast",
                "INVALID_POLICY",
                "single policies must not contain maxMulticast",
            )
    elif kind_valid and kind == "multi":
        if not max_present:
            _add_issue(
                issues,
                f"{path}/maxMulticast",
                "REQUIRED",
                "multi policies require maxMulticast",
            )
        else:
            max_multicast = portable_integer(value["maxMulticast"], minimum=1)
            if max_multicast is None or (
                allowed_routes_valid and max_multicast > len(allowed_routes)
            ):
                _add_issue(
                    issues,
                    f"{path}/maxMulticast",
                    "INVALID_POLICY",
                    "maxMulticast must be a safe integer in [1, allowedRoutes.length]",
                )

    return _PolicySnapshot(
        kind,
        allowed_routes,
        default_route,
        confidence if confidence_config_valid else None,
        confidence_present,
        max_multicast,
    )


def _result(
    reason_code: RouteSelectionReasonCode,
    request: _RequestSnapshot,
    selected_routes: tuple[str, ...],
    unknown_routes: tuple[str, ...],
    *,
    used_default: bool,
    escalated: bool,
) -> RouteSelectionResult:
    return RouteSelectionResult(
        routed=bool(selected_routes),
        reason_code=reason_code,
        requested_routes=tuple(request.requested_routes),
        selected_routes=tuple(selected_routes),
        unknown_routes=tuple(unknown_routes),
        confidence_basis_points=request.confidence_basis_points,
        used_default=used_default,
        escalated=escalated,
    )


def _evaluate(request_value: object, policy_value: object) -> RouteSelectionResult:
    issues: list[PrimitiveValidationIssue] = []
    request = _validate_request(request_value, issues)
    policy = _validate_policy(policy_value, issues)

    if policy.confidence_present and not request.confidence_present:
        _add_issue(
            issues,
            "#/request/confidenceBasisPoints",
            "CONFIDENCE_CONFIGURATION",
            "confidenceBasisPoints is required when policy confidence is configured",
        )
    elif not policy.confidence_present and request.confidence_present:
        _add_issue(
            issues,
            "#/request/confidenceBasisPoints",
            "CONFIDENCE_CONFIGURATION",
            "confidenceBasisPoints requires policy confidence configuration",
        )
    if issues:
        raise PrimitiveValidationError(
            tuple(issues),
            message="Route selection input is invalid",
        )

    allowed_set = frozenset(policy.allowed_routes)
    unknown_routes = tuple(
        route for route in request.requested_routes if route not in allowed_set
    )

    if (
        policy.confidence is not None
        and request.confidence_basis_points is not None
        and request.confidence_basis_points
        < policy.confidence.minimum_basis_points
    ):
        return _result(
            "ESCALATION_SELECTED_LOW_CONFIDENCE",
            request,
            (policy.confidence.escalation_route,),
            unknown_routes,
            used_default=False,
            escalated=True,
        )

    if not request.requested_routes:
        if policy.default_route is None:
            return _result(
                "NO_REQUESTED_ROUTE",
                request,
                (),
                (),
                used_default=False,
                escalated=False,
            )
        return _result(
            "DEFAULT_SELECTED_NO_REQUEST",
            request,
            (policy.default_route,),
            (),
            used_default=True,
            escalated=False,
        )

    if policy.kind == "single" and len(request.requested_routes) > 1:
        return _result(
            "MULTIPLE_ROUTES_FOR_SINGLE",
            request,
            (),
            unknown_routes,
            used_default=False,
            escalated=False,
        )
    if (
        policy.kind == "multi"
        and policy.max_multicast is not None
        and len(request.requested_routes) > policy.max_multicast
    ):
        return _result(
            "MULTICAST_LIMIT_EXCEEDED",
            request,
            (),
            unknown_routes,
            used_default=False,
            escalated=False,
        )

    if unknown_routes:
        if policy.default_route is None:
            return _result(
                "UNKNOWN_ROUTE",
                request,
                (),
                unknown_routes,
                used_default=False,
                escalated=False,
            )
        return _result(
            "DEFAULT_SELECTED_UNKNOWN_ROUTE",
            request,
            (policy.default_route,),
            unknown_routes,
            used_default=True,
            escalated=False,
        )

    requested_set = frozenset(request.requested_routes)
    selected_routes = tuple(
        route for route in policy.allowed_routes if route in requested_set
    )
    return _result(
        "REQUESTED_ROUTES_SELECTED",
        request,
        selected_routes,
        (),
        used_default=False,
        escalated=False,
    )


def evaluate_route_selection(
    request: object,
    policy: object,
) -> RouteSelectionResult:
    """Convert an untrusted route request into an auditable policy decision."""

    try:
        return _evaluate(request, policy)
    except PrimitiveValidationError:
        raise
    except Exception:
        raise PrimitiveValidationError(
            (
                PrimitiveValidationIssue(
                    "#",
                    "TYPE",
                    "route selection input could not be inspected safely",
                ),
            ),
            message="Route selection input is invalid",
        ) from None

File: packages/primitives/src/router.ts
import {
  PrimitiveValidationError,
  type PrimitiveValidationIssue,
  type PrimitiveValidationIssueCode,
} from "./errors.js";
import { compareUnicodeCodePoints } from "./json.js";
import type {
  RouteSelectionPolicy,
  RouteSelectionReasonCode,
  RouteSelectionRequest,
  RouteSelectionResult,
} from "./types.js";

const SAFE_ROUTE_ID = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;
const REQUEST_FIELDS = new Set(["requestedRoutes", "confidenceBasisPoints"]);
const POLICY_FIELDS = new Set([
  "kind",
  "allowedRoutes",
  "defaultRoute",
  "confidence",
  "maxMulticast",
]);
const CONFIDENCE_FIELDS = new Set(["minimumBasisPoints", "escalationRoute"]);

interface DataProperty {
  present: boolean;
  valid: boolean;
  value?: unknown;
}

interface RequestSnapshot {
  requestedRoutes: readonly string[];
  confidenceBasisPoints: number | null;
  confidencePresent: boolean;
}

interface ConfidenceSnapshot {
  minimumBasisPoints: number;
  escalationRoute: string;
}

interface PolicySnapshot {
  kind: "single" | "multi";
  allowedRoutes: readonly string[];
  defaultRoute: string | null;
  confidence: Readonly<ConfidenceSnapshot> | null;
  confidencePresent: boolean;
  maxMulticast: number | null;
}

function addIssue(
  issues: PrimitiveValidationIssue[],
  path: string,
  code: PrimitiveValidationIssueCode,
  message: string,
): void {
  issues.push({ path, code, message });
}

function pointerSegment(value: string): string {
  return value.replaceAll("~", "~0").replaceAll("/", "~1");
}

function isPlainRecord(value: unknown): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function unknownFields(
  value: Record<string, unknown>,
  allowed: ReadonlySet<string>,
  path: string,
  issues: PrimitiveValidationIssue[],
): void {
  for (const name of Object.getOwnPropertyNames(value).sort(compareUnicodeCodePoints)) {
    if (!allowed.has(name)) {
      addIssue(issues, `${path}/${pointerSegment(name)}`, "UNKNOWN_FIELD", "unknown field");
    }
  }
  if (Object.getOwnPropertySymbols(value).length > 0) {
    addIssue(issues, path, "UNKNOWN_FIELD", "symbol fields are not supported");
  }
}

function dataProperty(
  value: Record<string, unknown>,
  name: string,
  path: string,
  code: PrimitiveValidationIssueCode,
  issues: PrimitiveValidationIssue[],
): DataProperty {
  const descriptor = Object.getOwnPropertyDescriptor(value, name);
  if (descriptor === undefined) return { present: false, valid: false };
  if (!Object.hasOwn(descriptor, "value") || descriptor.enumerable !== true) {
    addIssue(issues, path, code, "expected an enumerable data field");
    return { present: true, valid: false };
  }
  return { present: true, valid: true, value: descriptor.value };
}

function validRouteId(value: unknown): value is string {
  return typeof value === "string" &&
    SAFE_ROUTE_ID.test(value) &&
    value !== "." &&
    value !== "..";
}

interface RouteArrayOptions {
  duplicateCode: "DUPLICATE_ID" | "DUPLICATE_SELECTION";
  requireNonEmpty: boolean;
  emptyCode: PrimitiveValidationIssueCode;
  emptyMessage: string;
}

function validateRouteArray(
  value: unknown,
  path: string,
  options: RouteArrayOptions,
  issues: PrimitiveValidationIssue[],
): readonly string[] {
  if (!Array.isArray(value)) {
    addIssue(issues, path, "TYPE", "expected a route ID array");
    return Object.freeze([]);
  }

  const lengthDescriptor = Object.getOwnPropertyDescriptor(value, "length");
  const length = lengthDescriptor !== undefined && Object.hasOwn(lengthDescriptor, "value")
    ? lengthDescriptor.value as number
    : 0;
  const indexNames: string[] = [];
  for (const name of Object.getOwnPropertyNames(value)
    .filter((item) => item !== "length")
    .sort(compareUnicodeCodePoints)) {
    const index = Number(name);
    if (!Number.isInteger(index) || index < 0 || index >= length || String(index) !== name) {
      addIssue(
        issues,
        `${path}/${pointerSegment(name)}`,
        "UNKNOWN_FIELD",
        "unknown array field",
      );
    } else {
      indexNames.push(name);
    }
  }
  if (Object.getOwnPropertySymbols(value).length > 0) {
    addIssue(issues, path, "UNKNOWN_FIELD", "symbol fields are not supported");
  }
  indexNames.sort((left, right) => Number(left) - Number(right));
  let expectedIndex = 0;
  for (const name of indexNames) {
    if (Number(name) !== expectedIndex) break;
    expectedIndex += 1;
  }
  if (expectedIndex !== length) {
    addIssue(issues, `${path}/${expectedIndex}`, "TYPE", "sparse route arrays are not supported");
  }
  if (options.requireNonEmpty && length === 0) {
    addIssue(issues, path, options.emptyCode, options.emptyMessage);
  }

  const routes: string[] = [];
  const seen = new Set<string>();
  for (const name of indexNames) {
    const descriptor = Object.getOwnPropertyDescriptor(value, name);
    const itemPath = `${path}/${name}`;
    if (
      descriptor === undefined ||
      !Object.hasOwn(descriptor, "value") ||
      descriptor.enumerable !== true
    ) {
      addIssue(issues, itemPath, "TYPE", "expected an enumerable route ID entry");
      continue;
    }
    if (!validRouteId(descriptor.value)) {
      addIssue(issues, itemPath, "UNSAFE_ID", "expected a safe route ID");
      continue;
    }
    const route = descriptor.value;
    if (seen.has(route)) {
      addIssue(
        issues,
        itemPath,
        options.duplicateCode,
        options.duplicateCode === "DUPLICATE_SELECTION"
          ? `duplicate requested route '${route}'`
          : `duplicate allowed route '${route}'`,
      );
    } else {
      seen.add(route);
    }
    routes.push(route);
  }
  return Object.freeze(routes);
}

function validateRequest(value: unknown, issues: PrimitiveValidationIssue[]): RequestSnapshot {
  const path = "#/request";
  if (!isPlainRecord(value)) {
    addIssue(issues, path, "TYPE", "expected a plain request object");
    return Object.freeze({
      requestedRoutes: Object.freeze([]),
      confidenceBasisPoints: null,
      confidencePresent: false,
    });
  }
  unknownFields(value, REQUEST_FIELDS, path, issues);

  const routesProperty = dataProperty(
    value,
    "requestedRoutes",
    `${path}/requestedRoutes`,
    "TYPE",
    issues,
  );
  let requestedRoutes: readonly string[] = Object.freeze([]);
  if (!routesProperty.present) {
    addIssue(issues, `${path}/requestedRoutes`, "REQUIRED", "requestedRoutes is required");
  } else if (routesProperty.valid) {
    requestedRoutes = validateRouteArray(
      routesProperty.value,
      `${path}/requestedRoutes`,
      {
        duplicateCode: "DUPLICATE_SELECTION",
        requireNonEmpty: false,
        emptyCode: "TYPE",
        emptyMessage: "",
      },
      issues,
    );
  }

  const confidenceProperty = dataProperty(
    value,
    "confidenceBasisPoints",
    `${path}/confidenceBasisPoints`,
    "INVALID_CONFIDENCE",
    issues,
  );
  let confidenceBasisPoints: number | null = null;
  if (confidenceProperty.present && confidenceProperty.valid) {
    if (
      !Number.isSafeInteger(confidenceProperty.value) ||
      (confidenceProperty.value as number) < 0 ||
      (confidenceProperty.value as number) > 10_000
    ) {
      addIssue(
        issues,
        `${path}/confidenceBasisPoints`,
        "INVALID_CONFIDENCE",
        "confidenceBasisPoints must be a safe integer in [0, 10000]",
      );
    } else {
      confidenceBasisPoints = confidenceProperty.value === 0 ? 0 : confidenceProperty.value as number;
    }
  }

  return Object.freeze({
    requestedRoutes,
    confidenceBasisPoints,
    confidencePresent: confidenceProperty.present,
  });
}

function validateConfidencePolicy(
  value: unknown,
  allowedRoutes: ReadonlySet<string>,
  allowedRoutesValid: boolean,
  issues: PrimitiveValidationIssue[],
): Readonly<ConfidenceSnapshot> | null {
  const path = "#/policy/confidence";
  if (!isPlainRecord(value)) {
    addIssue(issues, path, "INVALID_POLICY", "expected a plain confidence policy object");
    return null;
  }
  unknownFields(value, CONFIDENCE_FIELDS, path, issues);

  const minimumProperty = dataProperty(
    value,
    "minimumBasisPoints",
    `${path}/minimumBasisPoints`,
    "INVALID_CONFIDENCE",
    issues,
  );
  let minimumBasisPoints: number | undefined;
  if (!minimumProperty.present) {
    addIssue(issues, `${path}/minimumBasisPoints`, "REQUIRED", "minimumBasisPoints is required");
  } else if (minimumProperty.valid) {
    if (
      !Number.isSafeInteger(minimumProperty.value) ||
      (minimumProperty.value as number) < 1 ||
      (minimumProperty.value as number) > 10_000
    ) {
      addIssue(
        issues,
        `${path}/minimumBasisPoints`,
        "INVALID_CONFIDENCE",
        "minimumBasisPoints must be a safe integer in [1, 10000]",
      );
    } else {
      minimumBasisPoints = minimumProperty.value as number;
    }
  }

  const escalationProperty = dataProperty(
    value,
    "escalationRoute",
    `${path}/escalationRoute`,
    "INVALID_POLICY",
    issues,
  );
  let escalationRoute: string | undefined;
  if (!escalationProperty.present) {
    addIssue(issues, `${path}/escalationRoute`, "REQUIRED", "escalationRoute is required");
  } else if (escalationProperty.valid) {
    if (!validRouteId(escalationProperty.value)) {
      addIssue(issues, `${path}/escalationRoute`, "UNSAFE_ID", "expected a safe route ID");
    } else {
      escalationRoute = escalationProperty.value;
      if (allowedRoutesValid && !allowedRoutes.has(escalationRoute)) {
        addIssue(
          issues,
          `${path}/escalationRoute`,
          "INVALID_POLICY",
          "escalationRoute must be declared in allowedRoutes",
        );
      }
    }
  }

  if (minimumBasisPoints === undefined || escalationRoute === undefined) return null;
  return Object.freeze({ minimumBasisPoints, escalationRoute });
}

function validatePolicy(value: unknown, issues: PrimitiveValidationIssue[]): PolicySnapshot {
  const path = "#/policy";
  if (!isPlainRecord(value)) {
    addIssue(issues, path, "TYPE", "expected a plain policy object");
    return Object.freeze({
      kind: "single",
      allowedRoutes: Object.freeze([]),
      defaultRoute: null,
      confidence: null,
      confidencePresent: false,
      maxMulticast: null,
    });
  }
  unknownFields(value, POLICY_FIELDS, path, issues);

  const kindProperty = dataProperty(value, "kind", `${path}/kind`, "INVALID_POLICY", issues);
  let kind: "single" | "multi" = "single";
  let kindValid = false;
  if (!kindProperty.present) {
    addIssue(issues, `${path}/kind`, "REQUIRED", "policy kind is required");
  } else if (kindProperty.valid) {
    if (kindProperty.value !== "single" && kindProperty.value !== "multi") {
      addIssue(issues, `${path}/kind`, "INVALID_POLICY", "kind must be 'single' or 'multi'");
    } else {
      kind = kindProperty.value;
      kindValid = true;
    }
  }

  const routesProperty = dataProperty(
    value,
    "allowedRoutes",
    `${path}/allowedRoutes`,
    "TYPE",
    issues,
  );
  let allowedRoutes: readonly string[] = Object.freeze([]);
  const allowedIssuesStart = issues.length;
  if (!routesProperty.present) {
    addIssue(issues, `${path}/allowedRoutes`, "REQUIRED", "allowedRoutes is required");
  } else if (routesProperty.valid) {
    allowedRoutes = validateRouteArray(
      routesProperty.value,
      `${path}/allowedRoutes`,
      {
        duplicateCode: "DUPLICATE_ID",
        requireNonEmpty: true,
        emptyCode: "INVALID_POLICY",
        emptyMessage: "allowedRoutes must not be empty",
      },
      issues,
    );
  }
  const allowedRoutesValid = issues.length === allowedIssuesStart && routesProperty.valid;
  const allowedSet = new Set(allowedRoutes);

  const defaultProperty = dataProperty(
    value,
    "defaultRoute",
    `${path}/defaultRoute`,
    "INVALID_POLICY",
    issues,
  );
  let defaultRoute: string | null = null;
  if (defaultProperty.present && defaultProperty.valid) {
    if (!validRouteId(defaultProperty.value)) {
      addIssue(issues, `${path}/defaultRoute`, "UNSAFE_ID", "expected a safe route ID");
    } else {
      defaultRoute = defaultProperty.value;
      if (allowedRoutesValid && !allowedSet.has(defaultRoute)) {
        addIssue(
          issues,
          `${path}/defaultRoute`,
          "INVALID_POLICY",
          "defaultRoute must be declared in allowedRoutes",
        );
      }
    }
  }

  const confidenceProperty = dataProperty(
    value,
    "confidence",
    `${path}/confidence`,
    "INVALID_POLICY",
    issues,
  );
  let confidence: Readonly<ConfidenceSnapshot> | null = null;
  let confidenceConfigValid = false;
  if (confidenceProperty.present && confidenceProperty.valid) {
    const confidenceIssuesStart = issues.length;
    confidence = validateConfidencePolicy(
      confidenceProperty.value,
      allowedSet,
      allowedRoutesValid,
      issues,
    );
    confidenceConfigValid = confidence !== null && issues.length === confidenceIssuesStart;
  }

  const maxProperty = dataProperty(
    value,
    "maxMulticast",
    `${path}/maxMulticast`,
    "INVALID_POLICY",
    issues,
  );
  let maxMulticast: number | null = null;
  if (kindValid && kind === "single") {
    if (maxProperty.present) {
      addIssue(
        issues,
        `${path}/maxMulticast`,
        "INVALID_POLICY",
        "single policies must not contain maxMulticast",
      );
    }
  } else if (kindValid && kind === "multi") {
    if (!maxProperty.present) {
      addIssue(issues, `${path}/maxMulticast`, "REQUIRED", "multi policies require maxMulticast");
    } else if (maxProperty.valid) {
      if (
        !Number.isSafeInteger(maxProperty.value) ||
        (maxProperty.value as number) < 1 ||
        (allowedRoutesValid && (maxProperty.value as number) > allowedRoutes.length)
      ) {
        addIssue(
          issues,
          `${path}/maxMulticast`,
          "INVALID_POLICY",
          "maxMulticast must be a safe integer in [1, allowedRoutes.length]",
        );
      } else {
        maxMulticast = maxProperty.value as number;
      }
    }
  }

  return Object.freeze({
    kind,
    allowedRoutes,
    defaultRoute,
    confidence: confidenceConfigValid ? confidence : null,
    confidencePresent: confidenceProperty.present,
    maxMulticast,
  });
}

function result(
  reasonCode: RouteSelectionReasonCode,
  request: RequestSnapshot,
  selectedRoutes: readonly string[],
  unknownRoutes: readonly string[],
  usedDefault: boolean,
  escalated: boolean,
): RouteSelectionResult {
  const selectedSnapshot = Object.freeze([...selectedRoutes]);
  const unknownSnapshot = Object.freeze([...unknownRoutes]);
  const output: RouteSelectionResult = {
    routed: selectedSnapshot.length > 0,
    reasonCode,
    requestedRoutes: Object.freeze([...request.requestedRoutes]),
    selectedRoutes: selectedSnapshot,
    unknownRoutes: unknownSnapshot,
    confidenceBasisPoints: request.confidenceBasisPoints,
    usedDefault,
    escalated,
  };
  return Object.freeze(output);
}

function evaluate(requestValue: unknown, policyValue: unknown): RouteSelectionResult {
  const issues: PrimitiveValidationIssue[] = [];
  const request = validateRequest(requestValue, issues);
  const policy = validatePolicy(policyValue, issues);

  if (policy.confidencePresent && !request.confidencePresent) {
    addIssue(
      issues,
      "#/request/confidenceBasisPoints",
      "CONFIDENCE_CONFIGURATION",
      "confidenceBasisPoints is required when policy confidence is configured",
    );
  } else if (!policy.confidencePresent && request.confidencePresent) {
    addIssue(
      issues,
      "#/request/confidenceBasisPoints",
      "CONFIDENCE_CONFIGURATION",
      "confidenceBasisPoints requires policy confidence configuration",
    );
  }
  if (issues.length > 0) {
    throw new PrimitiveValidationError(issues, "Route selection input is invalid");
  }

  const allowedSet = new Set(policy.allowedRoutes);
  const unknownRoutes = request.requestedRoutes.filter((route) => !allowedSet.has(route));

  if (
    policy.confidence !== null &&
    request.confidenceBasisPoints !== null &&
    request.confidenceBasisPoints < policy.confidence.minimumBasisPoints
  ) {
    return result(
      "ESCALATION_SELECTED_LOW_CONFIDENCE",
      request,
      [policy.confidence.escalationRoute],
      unknownRoutes,
      false,
      true,
    );
  }

  if (request.requestedRoutes.length === 0) {
    return policy.defaultRoute === null
      ? result("NO_REQUESTED_ROUTE", request, [], [], false, false)
      : result("DEFAULT_SELECTED_NO_REQUEST", request, [policy.defaultRoute], [], true, false);
  }

  if (policy.kind === "single" && request.requestedRoutes.length > 1) {
    return result("MULTIPLE_ROUTES_FOR_SINGLE", request, [], unknownRoutes, false, false);
  }
  if (
    policy.kind === "multi" &&
    policy.maxMulticast !== null &&
    request.requestedRoutes.length > policy.maxMulticast
  ) {
    return result("MULTICAST_LIMIT_EXCEEDED", request, [], unknownRoutes, false, false);
  }

  if (unknownRoutes.length > 0) {
    return policy.defaultRoute === null
      ? result("UNKNOWN_ROUTE", request, [], unknownRoutes, false, false)
      : result(
        "DEFAULT_SELECTED_UNKNOWN_ROUTE",
        request,
        [policy.defaultRoute],
        unknownRoutes,
        true,
        false,
      );
  }

  const requestedSet = new Set(request.requestedRoutes);
  const selectedRoutes = policy.allowedRoutes.filter((route) => requestedSet.has(route));
  return result("REQUESTED_ROUTES_SELECTED", request, selectedRoutes, [], false, false);
}

/**
 * Deterministically convert a validated route request into an auditable
 * selection. State precedence is low confidence, empty request, count limits,
 * unknown/default handling, then canonical selection.
 */
export function evaluateRouteSelection(
  request: RouteSelectionRequest,
  policy: RouteSelectionPolicy,
): RouteSelectionResult {
  try {
    return evaluate(request, policy);
  } catch (error) {
    if (error instanceof PrimitiveValidationError) throw error;
    throw new PrimitiveValidationError(
      [{ path: "#", code: "TYPE", message: "route selection input could not be inspected safely" }],
      "Route selection input is invalid",
    );
  }
}
